Create JobHookReceiver subclass of Job (#2056)
nautobot/nautobot: nautobot/docs/additional-features/job-hooks.md

> # Job hooks

Reviewer: ```suggestion
# Job Hooks
```

> A job hook is a mechanism for automatically starting a [job](./jobs.md) when an object is changed. Job hooks are similar to [webhooks](../models/extras/webhook.md) except change events initiate a `JobHookReceiver` job instead of a web request. Job hooks are configured in the web UI under Jobs > Job Hooks.

Reviewer: ```suggestion
A Job Hook is a mechanism for automatically starting a [job](./jobs.md) when an object is changed. Job Hooks are similar to [webhooks](../models/extras/webhook.md) except that an object change event initiates a `JobHookReceiver` job instead of a web request. Job hooks are configured in the web UI under **Jobs > Job Hooks**.
```

> ## Job hook receivers

Reviewer: ```suggestion
## Job Hook Receivers
```

> Job hooks are only able to initiate a specific type of job called a job hook receiver. These are jobs that subclass the `nautobot.extras.jobs.JobHookReceiver` class. Job hook receivers are similar to normal jobs except they are hard coded to accept only an `object_change` [variable](jobs.md#variables). Job hook receivers are hidden from the jobs listing UI by default but function similarly to other jobs. The job hook receiver class only implements one method called `receive_jobhook`.

Reviewer: ```suggestion
Job Hooks are only able to initiate a specific type of job called a **Job Hook Receiver**. These are jobs that subclass the `nautobot.extras.jobs.JobHookReceiver` class. Job hook receivers are similar to normal jobs except they are hard coded to accept only an `object_change` [variable](jobs.md#variables). Job Hook Receivers are hidden from the jobs listing UI by default but otherwise function similarly to other jobs. The `JobHookReceiver` class only implements one method called `receive_jobhook`.
```

!!! important

>     To prevent negatively impacting system performance through an infinite loop, a change that was made by a job hook receiver job will not trigger another job hook receiver job to run.

Reviewer: ```suggestion
    To prevent negatively impacting system performance through an infinite loop, a change that was made by a `JobHookReceiver` job will not trigger another `JobHookReceiver` job to run.
```

> ### Example job hook receiver

Reviewer: ```suggestion
### Example Job Hook Receiver
```

```py
from nautobot.extras.choices import ObjectChangeActionChoices
from nautobot.extras.jobs import JobHookReceiver

> class ExampleJobHookReceiver(JobHookReceiver):

Reviewer: ```suggestion

class ExampleJobHookReceiver(JobHookReceiver):
```

        # return on delete action
        if action == ObjectChangeActionChoices.ACTION_DELETE:
            return

        # log diff output
        snapshots = change.get_snapshots()
        self.log_info(f"DIFF: {snapshots['differences']}")

        # validate changes to serial field
        if "serial" in snapshots["differences"]["added"]:
            old_serial = snapshots["differences"]["removed"]["serial"]
            new_serial = snapshots["differences"]["added"]["serial"]
            self.log_info(f"{changed_object} serial has been changed from {old_serial} to {new_serial}")

            # Check the new serial is valid and revert if necessary
            if not self.validate_serial(new_serial):
                changed_object.serial = old_serial
                changed_object.save()
                self.log_info(f"{changed_object} serial {new_serial} was not valid. Reverted to {old_serial}")

            self.log_success(message=f"Serial validation completed for {changed_object}")

    def validate_serial(self, serial):
        # add business logic to validate serial
        return False
```

### The `receive_jobhook()` Method

> All job hook receivers must implement a `receive_jobhook()` method. This method accepts three arguments:

Reviewer: ```suggestion
All `JobHookReceiver` subclasses must implement a `receive_jobhook()` method. This method accepts three arguments:
```

# Job Hooks

A Job Hook is a mechanism for automatically starting a [job](./jobs.md) when an object is changed. Job Hooks are similar to [webhooks](../models/extras/webhook.md) except that an object change event initiates a `JobHookReceiver` job instead of a web request. Job hooks are configured in the web UI under **Jobs > Job Hooks**.

## Job Hook Receivers

> Job Hooks are only able to initiate a specific type of job called a **Job Hook Receiver**. These are jobs that subclass the `nautobot.extras.jobs.JobHookReceiver` class. Job hook receivers are similar to normal jobs except they are hard coded to accept only an `object_change` [variable](jobs.md#variables). Job Hook Receivers are hidden from the jobs listing UI by default but otherwise function similarly to other jobs. The `JobHookReceiver` class only implements one method called `receive_jobhook`.

Reviewer 2: nitpick: perhaps change to `receive_job_hook` throughout?
